Heavily refactored Random level generation.

ptrckkrmr opened the pull request:
This removes the near-duplication in the methods named
FindPathNorthEast, FindPathNorthWest, FindPathSouthEast and
FindPathSouthWest, replacing these with a single method named
FindPathSegment.

Also, the path generation should be extensible, allowing for a
custom amount of iterations before completing the path.

However, this has not been tested yet.

ptrckkrmr committed:
Heavily refactored Random level generation.

This removes the near-duplication in the methods named
FindPathNorthEast, FindPathNorthWest, FindPathSouthEast and
FindPathSouthWest, replacing these with a single method named
FindPathSegment.

Also, the path generation should be extensible, allowing for a
custom amount of iterations before completing the path.

However, this has not been tested yet.

TRvanRossum commented:
Played around with this a bit: right now the generated levels do not include a laser, and I'm uncertain if they are always solvable, especially considering walls are still randomly rotated during the rendering of the map, and thus this mechanic could make a map unsolvable. I'll look more into it tomorrow.

ptrckkrmr committed:
Fixed Emitter and Target not being added properly.

Also fixed some issues in random offsets: these could sometimes
be outside the active graph.

ptrckkrmr commented:
Missing laser is fixed now. The levels should always be solvable with at most `2n - 1` mirrors after `n` iterations, but please do check tomorrow. I might have missed something.

ptrckkrmr committed:
Changed SquareGraph constructor to throw ArgumentOutOfRangeException

This is a more specific exception than the general ArgumentException,
and gives a better indication of the cause of the Exception.

ptrckkrmr committed:
Added some more boundary tests for SquareGraph.

Also renamed test cases to more accurately reflect their purpose.
The general format is now: Test[NameOfMethod][DescriptionOfTestCase].

ptrckkrmr committed:
Added script that updates the VS project to use .NET 4

This solves the warning about Meta needing mscorlib version 4.0.0.0,
which was due to Unity generating VS files for .NET 3.5

The script is derived from the GitHub Gist
https://gist.github.com/MattRix/daf67d66227c61501397

ptrckkrmr committed:
Fixed StyleCop issues in refactored RandomLevel classes.

TRvanRossum commented:
Taken another look: the code seems to construct a critical path from the emitter to the target, before adding walls, and walls are never put on the critical path. That should be enough to guarantee that the map is solvable. Going to merge this.
